diff --git a/src/core/ngx_connection.h b/src/core/ngx_connection.h
index e4dfe58..9250982 100644
--- a/src/core/ngx_connection.h
+++ b/src/core/ngx_connection.h
@@ -133,6 +133,7 @@ struct ngx_connection_s {
     ngx_listening_t    *listening;
 
     off_t               sent;
+    off_t               received;
 
     ngx_log_t          *log;
 
@@ -189,6 +190,9 @@ struct ngx_connection_s {
 #if (NGX_THREADS || NGX_COMPAT)
     ngx_thread_task_t  *sendfile_task;
 #endif
+	 
+	time_t start_sec;
+    ngx_msec_t start_msec;
 };
 
 
diff --git a/src/http/ngx_http_request.c b/src/http/ngx_http_request.c
index c88c271..dbe601f 100644
--- a/src/http/ngx_http_request.c
+++ b/src/http/ngx_http_request.c
@@ -225,6 +225,11 @@ ngx_http_init_connection(ngx_connection_t *c)
 
     c->data = hc;
 
+	// 连接开始的地儿
+	ngx_time_t *tp = ngx_timeofday();
+    c->start_sec = tp->sec;
+    c->start_msec = tp->msec;
+	
     /* find the server configuration for the address:port */
 
     port = c->listening->servers;
@@ -474,6 +479,7 @@ ngx_http_wait_request_handler(ngx_event_t *rev)
     }
 
     b->last += n;
+    c->received += n;
 
     if (hc->proxy_protocol) {
         hc->proxy_protocol = 0;
@@ -1451,7 +1465,8 @@ ngx_http_read_request_header(ngx_http_request_t *r)
     }
 
     r->header_in->last += n;
-
+    c->received += n;
+ 
     return n;
 }
 
@@ -2992,6 +3007,8 @@ ngx_http_set_keepalive(ngx_http_request_t *r)
 
         c->data = r;
 
+        /* bytes in the buffer have already been counted */
+        c->received = 0;
         c->sent = 0;
         c->destroyed = 0;
 
@@ -3230,6 +3247,7 @@ ngx_http_keepalive_handler(ngx_event_t *rev)
     }
 
     c->sent = 0;
+    c->received = n;
     c->destroyed = 0;
 
     ngx_del_timer(rev);
@@ -3321,6 +3339,7 @@ ngx_http_lingering_close_handler(ngx_event_t *rev)
             return;
         }
 
+	c->received += n;
     } while (rev->ready);
 
     if (ngx_handle_read_event(rev, 0) != NGX_OK) {
diff --git a/src/http/ngx_http_write_filter_module.c b/src/http/ngx_http_write_filter_module.c
index 0036231..f3d686b 100644
--- a/src/http/ngx_http_write_filter_module.c
+++ b/src/http/ngx_http_write_filter_module.c
@@ -12,6 +12,7 @@
 
 static ngx_int_t ngx_http_write_filter_init(ngx_conf_t *cf);
 
+ngx_int_t (*ngx_http_write_filter_stat)(ngx_http_request_t *r) = NULL;
 
 static ngx_http_module_t  ngx_http_write_filter_module_ctx = {
     NULL,                                  /* preconfiguration */
@@ -253,6 +254,12 @@ ngx_http_write_filter(ngx_http_request_t *r, ngx_chain_t *in)
 
     chain = c->send_chain(c, r->out, limit);
 
+	if (ngx_http_write_filter_stat != NULL) {
+        if (ngx_http_write_filter_stat(r) == NGX_ERROR) {
+            return NGX_ERROR;
+        }
+    }
+	
     ngx_log_debug1(NGX_LOG_DEBUG_HTTP, c->log, 0,
                    "http write filter %p", chain);
 
